chapter_reinforcement_learning/value-iter.md

```diff
@@ -0,0 +1,152 @@
+$$V_{k+1}(s) = \sum_{a \in \mathcal{A}} \pi(a \mid s) \Big[ r(s,  a) + \gamma\  \sum_{s' \in \mathcal{S}} P(s' \mid s, a) V_k(s') \Big];\ \text{for all } s \in \mathcal{S}.$$
+
+It turns our that as $k \to \infty$ the value function estimated by the Value Iteration algorithm converges to the optimal value function irrespective of the initialization $V_0$,
+$$V^*(s) = \lim_{k \to \infty} V_k(s);\ \text{for all states } s \in \mathcal{S}.$$
+
+The same value iteration can be equivalently written using the action-value function as
+$$Q_{k+1}(s, a) = r(s, a) + \gamma \sum_{s' \in \mathcal{S}} P(s' \mid s, a) \sum_{a' \in \mathcal{A}} \pi(a' \mid s')\ Q_k (s', a'); \text{ for all } s \in \mathcal{S}, a \in \mathcal{A}.$$
+
+In this case we initialize $Q_0(s, a)$ to some arbitrary value for all $s \in \mathcal{S}$ and $a \in \mathcal{A}$. Again we have $Q^*(s, a) = \lim_{k \to \infty} Q_k(s, a)$ for all $s \in \mathcal{S}$ and $a \in \mathcal{A}$.
+
+
+## Implementation of Value Iteration
+
+We next show how to implement Value Iteration for a navigation problem called FrozenLake from [Open AI Gym](https://gym.openai.com). We first need to setup the enviroment as shown in the following code.
+```{.python .input}
+#@tab all
+
+%matplotlib inline
+%pip install gym==0.21.0
+import numpy as np
+import random
+from d2l import torch as d2l
+
+seed = 0       # random number generator seed
+gamma = 0.95   # discount factor
+num_iters = 10 # number of iterations
+random.seed(seed) # set the random seed to ensure results can be reproduced
+np.random.seed(seed)
+
+# now set up the environment
+env_info = d2l.make_env('FrozenLake-v1')
+```
+
+In the FrozenLake environment, a robot/agent moves on a $4 \times 4$ grid (these are the states) with actions that are "up" ($\uparrow$), "down" ($\rightarrow$), "left" ($\leftarrow$), and "right" ($\rightarrow$). The environment contains a number of holes (H) cells and forzen (F) cells as well as a goal (G), all of which are unknown to the robot. To keep the problem simple,
+we assume the robot has reliable actions, i.e. $P(s' \mid s, a) = 1$ for all $s \in \mathcal{S}, a \in \mathcal{A}$. If the robot reaches the goal, the trial ends and the robot receives a reward of 1 irrespective of the action. The reward at any other state is zero for all actions. The objective of the robot is to learn a policy that reaches the goal location (G) from a given start location (S) (this is $s_0$) to maximize the return.
+
+The following function implements Value Iteration:
+
+```{.python .input}
+#@tab all
+
+
+def value_iteration(einfo, gamma, num_iters):
+    '''
+        gamma: discount factor
+        num_iters: number of iteration
+        env_info: contains MDP and environment related information
+    '''
+    env_desc      = einfo['desc'] # 2D array specifying what each grid item means
```

Comment:
This line exceeds the maximum line length 78, you can check the length by enabling the ruler extension. please refer to http://book.d2l.ai/user/code.html for more details.